RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -217,5 +219,101 @@ int64_t make_right_orthogonalizer(
+ *      of A_hat = V diag(eigvals) V'.
+ * @param[in,out] k
+ *      On entry: the number of columns to be used in V for the first iteration.
+ *      On exit: the number of columns in V.
+ * @param[in] mu_min
+ *      The smallest value of mu for which we want (V, eigvals) to define a useful
+ *      preconditioner for regularized linear systems (A + mu*I)x = b.
+ * @param[in] state
+ *      The RNGState used for random sketching inside the algorithm.
+ * 
+ * @param[in] num_syps_passes
+ *      A very small nonnegative integer. Optional; defaults to 3. 
+ *      This controls the number of power iterations used in computing a sketch of A.
+ *      Increasing the value would reduce |A - V diag(eigvals) V'|| when the number
+ *      of columns in V is fixed.
+ * @param[in] num_steps_power_iter_error_est
+ *      A small positive integer. Optional; defaults to 10.
+ *      This controls the number of power iterations used to estimate the spectral
+ *      norm of A - V diag(eigvals) V'. If this value is too small then (V, eigvals)
+ *      might not lead to good preconditioners for A+mu*I even when mu is >= mu_min.
+ * 
+ * @returns
+ *      An RNGState that the calling function should use the next
+ *      time it needs an RNGState.
+ */
+template <typename T, typename RNG>
+RandBLAS::RNGState<RNG> nystrom_pc_data(
+    blas::Uplo uplo,
+    const T* A,
+    int64_t m,
+    std::vector<T> &V,
+    std::vector<T> &eigvals,
+    int64_t &k,
+    T mu_min,
+    RandBLAS::RNGState<RNG> state,
+    int64_t num_syps_passes = 3,
+    int64_t num_steps_power_iter_error_est = 10
+) {
+    RandLAPACK::SYPS<T, RNG> SYPS(num_syps_passes, 1, false, false);
+    // ^ Define a symmetric power sketch algorithm.
+    //      (*) Stabilize power iteration with pivoted-LU after every
+    //          mulitplication with A.
+    //      (*) Do not check condition numbers or log to std::out.
+    RandLAPACK::HQRQ<T> Orth(false, false); 
+    // ^ Define an orthogonalizer for a symmetric rangefinder.
+    //      (*) Get a dense representation of Q from Householder QR.
+    //      (*) Do not check condition numbers or log to std::out.
+    RandLAPACK::SYRF<T, RNG> SYRF(SYPS, Orth, false, false);
+    // ^ Define the symmetric rangefinder algorithm.
+    //      (*) Use power sketching followed by Householder orthogonalization.
+    //      (*) Do not check condition numbers or log to std::out.
+    RandLAPACK::REVD2<T, RNG> NystromAlg(SYRF, num_steps_power_iter_error_est, false);
+    // ^ Define the algorithm for low-rank approximation via Nystrom.
+    //      (*) Handle accuracy requests by estimating ||A - V diag(eigvals) V'||
+    //          with "num_steps_power_iter_error_est" steps of power iteration.
+    //      (*) Do not log to std::out.
+    T tol = mu_min / 5;
+    // ^ Set tolerance to something materially smaller than the smallest
+    //   regularization parameter the user claims to need.
```

Comment:
I don't have quite enough context to know if this makes sense, but `/ 5` is materially smaller?

Reply:
It is indeed!